RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithmBase.cc

```diff
@@ -0,0 +1,106 @@
+}
+
+void VectorHitBuilderAlgorithmBase::printClusters(const edmNew::DetSetVector<Phase2TrackerCluster1D>& clusters) const {
+  int nCluster = 0;
+  int numberOfDSV = 0;
+  for (const auto& DSViter : clusters) {
+    ++numberOfDSV;
+    // Loop over the clusters in the detector unit
+    for (const auto& clustIt : DSViter) {
+      nCluster++;
+      // get the detector unit's id
+      const GeomDetUnit* geomDetUnit(theTkGeom->idToDetUnit(DSViter.detId()));
+      if (!geomDetUnit)
+        return;
+      printCluster(geomDetUnit, &clustIt);
+    }
+  }
+  LogDebug("VectorHitBuilder") << " Number of input clusters: " << nCluster << std::endl;
+}
+
+void VectorHitBuilderAlgorithmBase::printCluster(const GeomDet* geomDetUnit,
+                                                 const Phase2TrackerCluster1D* clustIt) const {
+  if (!geomDetUnit)
+    return;
+  const PixelGeomDetUnit* pixelGeomDetUnit = dynamic_cast<const PixelGeomDetUnit*>(geomDetUnit);
+  const PixelTopology& topol = pixelGeomDetUnit->specificTopology();
+  if (!pixelGeomDetUnit)
+    return;
+
+  unsigned int layer = theTkTopo->layer(geomDetUnit->geographicalId());
+  unsigned int module = theTkTopo->module(geomDetUnit->geographicalId());
+  LogTrace("VectorHitBuilder") << "Layer:" << layer << " and DetId: " << geomDetUnit->geographicalId().rawId()
+                               << std::endl;
+  TrackerGeometry::ModuleType mType = theTkGeom->getDetectorType(geomDetUnit->geographicalId());
+  if (mType == TrackerGeometry::ModuleType::Ph2PSP)
+    LogTrace("VectorHitBuilder") << "Pixel cluster (module:" << module << ") " << std::endl;
+  else if (mType == TrackerGeometry::ModuleType::Ph2SS || mType == TrackerGeometry::ModuleType::Ph2PSS)
+    LogTrace("VectorHitBuilder") << "Strip cluster (module:" << module << ") " << std::endl;
+  else
+    LogTrace("VectorHitBuilder") << "no module?!" << std::endl;
+  LogTrace("VectorHitBuilder") << "with pitch:" << topol.pitch().first << " , " << topol.pitch().second << std::endl;
+  LogTrace("VectorHitBuilder") << " and width:" << pixelGeomDetUnit->surface().bounds().width()
+                               << " , lenght:" << pixelGeomDetUnit->surface().bounds().length() << std::endl;
+
+  auto&& lparams = theCpe->localParameters(*clustIt, *pixelGeomDetUnit);
+  Global3DPoint gparams = pixelGeomDetUnit->surface().toGlobal(lparams.first);
+
+  LogTrace("VectorHitBuilder") << "\t global pos " << gparams << std::endl;
+  LogTrace("VectorHitBuilder") << "\t local  pos " << lparams.first << "with err " << lparams.second << std::endl;
+  LogTrace("VectorHitBuilder") << std::endl;
+
+  return;
+}
+
+void VectorHitBuilderAlgorithmBase::loadDetSetVector(std::unordered_map<DetId, std::vector<VectorHit> >& theMap,
+                                                     edmNew::DetSetVector<VectorHit>& theCollection) const {
+  for (const auto& it : theMap) {
+    edmNew::DetSetVector<VectorHit>::FastFiller vh_col(theCollection, it.first);
```

Comment:
Thanks, done!

Reply:
Mhh, this actually leads to an extreme increase in run time. I haven't let an event run to completion, but the lower bound is 10 minutes for one event. So far I haven't understood why it would behave like this.

Reply:
how did the memory look for that job? I'm curious if something went off and it started swapping

Reply:
uhm, I guess .reserve works worse than the auto-increasing capacity during push_back.

In this case one-pass loop (or use if known from before) to compute the total size should be applied to `theCollection.reserve(theMap.size(), totalSize);` outside of this loop

Reply:
I'll try that (probably on Monday). Memory is behaving fine, for what its worth.

Reply:
I was guessing a bug when I asked about memory, but then after thinking that 13K small .reserve which themselves can trigger reallocations can kill the performance much worse than the default realloc from a push_back which I think goes in multiples of 2.

Reply:
I have implemented the suggested `reserve` statements, VHs are now counted when the containers are filled, so no need for an additional loop.